File: Samples/IoT-GPIO/js/js/scenario1-getandsetpin.js
//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved
//// This code is licensed under the MIT License (MIT).

(function () {
    "use strict";

    var pin;

    var scenarioControls;
    var currentPinValue;

    var page = WinJS.UI.Pages.define("/html/scenario1-getandsetpin.html", {
        ready: function (element, options) {
            document.getElementById("startStopButton").addEventListener("click", startStopScenario);
            document.getElementById("setHighButton").addEventListener("click", setPinHigh);
            document.getElementById("setLowButton").addEventListener("click", setPinLow);
            document.getElementById("getValueButton").addEventListener("click", getPinValue);

            scenarioControls = document.getElementById("scenarioControls");
            currentPinValue = document.getElementById("currentPinValue");
        },

        unload: function () {
            stopScenario();
        }
    });

    function startScenario() {
        var gpioController = Windows.Devices.Gpio.GpioController.getDefault();

        // Set up our GPIO pin for setting values.
        // If this next line crashes with "Unable to get property 'openPin'
        // of undefined or null reference", then the problem is that there
        // is no GPIO controller on the device.
        pin = gpioController.openPin(5);

        // Configure pin for output.
        pin.setDriveMode(Windows.Devices.Gpio.GpioPinDriveMode.output);
    }

    function stopScenario() {
        // Release the GPIO pin.
        if (pin) {
            pin.close();
            pin = null;
        }
    }

    function startStopScenario(e) {
        if (pin) {
            stopScenario();
            e.srcElement.innerText = "Start";
            scenarioControls.style.display = "none";
        } else {
            startScenario();
            e.srcElement.innerText = "Stop";
            scenarioControls.style.display = "block";
        }
    }

    function setPinHigh() {
        // Set the pin value to High.
        pin.write(Windows.Devices.Gpio.GpioPinValue.high);
    }

    function setPinLow() {
        // Set the pin value to Low.
        pin.write(Windows.Devices.Gpio.GpioPinValue.low);
    }

    function getPinValue() {
        // Change the GUI to reflect the current pin value.
        if (pin.read() == Windows.Devices.Gpio.GpioPinValue.low) {
            currentPinValue.innerText = "Low";
        } else {
            currentPinValue.innerText = "High";
        }
    }

})();

File: Samples/IoT-GPIO/cppwinrt/Scenario1_GetAndSetPin.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario1_GetAndSetPin.h"
#include "Scenario1_GetAndSetPin.g.cpp"
#include "SampleConfiguration.h"

using namespace winrt;
using namespace winrt::Windows::Devices::Gpio;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::UI::Xaml;

namespace winrt::SDKTemplate::implementation
{
    Scenario1_GetAndSetPin::Scenario1_GetAndSetPin()
    {
        InitializeComponent();
    }

    void Scenario1_GetAndSetPin::OnNavigatedFrom(Windows::UI::Xaml::Navigation::NavigationEventArgs const&)
    {
        StopScenario();
    }

    void Scenario1_GetAndSetPin::StartScenario()
    {
        GpioController gpio = GpioController::GetDefault();

        // Set up our GPIO pin for setting values.
        // If this next line crashes with a NullReferenceException,
        // then the problem is that there is no GPIO controller on the device.
        pin = gpio.OpenPin(LED_PIN);

        // Configure pin for output.
        pin.SetDriveMode(GpioPinDriveMode::Output);
    }

    void Scenario1_GetAndSetPin::StopScenario()
    {
        // Release the GPIO pin.
        if (pin != nullptr)
        {
            pin.Close();
            pin = nullptr;
        }
    }

    void Scenario1_GetAndSetPin::StartStopScenario_Click(IInspectable const&, RoutedEventArgs const&)
    {
        if (pin != nullptr)
        {
            StopScenario();
            StartStopButton().Content(box_value(L"Start"));
            ScenarioControls().Visibility(Visibility::Collapsed);
        }
        else
        {
            StartScenario();
            StartStopButton().Content(box_value(L"Stop"));
            ScenarioControls().Visibility(Visibility::Visible);
        }
    }

    void Scenario1_GetAndSetPin::SetPinHigh_Click(IInspectable const&, RoutedEventArgs const&)
    {
        // Set the pin value to High.
        pin.Write(GpioPinValue::High);
    }

    void Scenario1_GetAndSetPin::SetPinLow_Click(IInspectable const&, RoutedEventArgs const&)
    {
        // Set the pin value to Low.
        pin.Write(GpioPinValue::Low);
    }

    void Scenario1_GetAndSetPin::GetPinValue_Click(IInspectable const&, RoutedEventArgs const&)
    {
        // Change the GUI to reflect the current pin value.
        CurrentPinValue().Text(PinValueToString(pin.Read()));
    }

}
